### PR TITLE
Preemption within ClusterQueue and cohort

### LINKED ISSUE #83: Support for workload preemption
Preemption can be useful to reclaim borrowed capacity, however the obvious tradeoff is interrupting workloads and potentially losing significant progress.

There are two high-level design decision we need to make and whether they should be tunable:
1. What triggers preemption? reclaiming borrowed capacity? workload priority?
2. What is the scope? is preemption is cohort knob? a capacity knob? a queue knob?

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind feature

/kind api-change

#### What this PR does / why we need it:

Implements https://github.com/kubernetes-sigs/kueue/tree/05da320/keps/83-workload-preemption

#### Which issue(s) this PR fixes:

Fixes #83

### CODE REVIEW

File: pkg/scheduler/preemption/preemption_test.go [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+			}).
+			Obj(),
+	}
+	cases := map[string]struct {
+		admitted      []kueue.Workload
+		incoming      *kueue.Workload
+		targetCQ      string
+		assignment    flavorassigner.Assignment
+		wantPreempted sets.Set[string]
+	}{
```
Comment (ahg-g) [inline pkg/scheduler/preemption/preemption_test.go:90]: A suggestion for a weird test: Is it worth it to have a test where a workload has two podSets, each one require preemption for the same resource, but they get different flavors after preemption?

File: test/integration/scheduler/preemption_test.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+			q = testing.MakeLocalQueue("q", ns.Name).ClusterQueue(cq.Name).Obj()
+			gomega.Expect(k8sClient.Create(ctx, q)).To(gomega.Succeed())
+		})
+
+		ginkgo.AfterEach(func() {
+			gomega.Expect(util.DeleteWorkloadsInNamespace(ctx, k8sClient, ns)).To(gomega.Succeed())
+			util.ExpectClusterQueueToBeDeleted(ctx, k8sClient, cq, true)
+		})
+
+		ginkgo.It("Should preempt Workloads with lower priority when there is not enough quota", func() {
```
Comment (ahg-g) [inline test/integration/scheduler/preemption_test.go:85]: a test with node affinity and/or tolerations to verify that those are taken into account when selecting the flavors on preemption.

**Review by ahg-g — COMMENTED** [review]
The suggestions for optimizations can be done on followup PRs.

File: pkg/scheduler/preemption/preemption.go [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/constants"
+	"sigs.k8s.io/kueue/pkg/scheduler/flavorassigner"
+	"sigs.k8s.io/kueue/pkg/util/priority"
+	"sigs.k8s.io/kueue/pkg/util/routine"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+const parallelPreemptions = 8
```
Comment (tenzen-y) [inline pkg/scheduler/preemption/preemption.go:43]: Do you have reasons why this is hardcoded?

File: apis/kueue/v1alpha2/clusterqueue_types.go [hunk 4]
```diff
@@ -272,6 +289,48 @@ type Usage struct {
 	Borrowed *resource.Quantity `json:"borrowing,omitempty"`
 }
 
+type PreemptionPolicy string
+
+const (
+	PreemptionPolicyNever                    PreemptionPolicy = "Never"
+	PreemptionPolicyReclaimFromLowerPriority PreemptionPolicy = "ReclaimFromLowerPriority"
+	PreemptionPolicyReclaimFromAny           PreemptionPolicy = "ReclaimFromAny"
+	PreemptionPolicyLowerPriority            PreemptionPolicy = "LowerPriority"
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/kueue/v1alpha2/clusterqueue_types.go:298]: The word "Reclaim" might be necessary if we ever add "stealing": when a ClusterQueue is able to preempt while borrowing quota for high priority workloads.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/kueue/v1alpha2/clusterqueue_types.go:298]: An alternative:

```yaml
preemption:
  reclaimWithinCohort: Never|LowerPriority|Any
  withinClusterQueue: Never|LowerPriority
```

[the remark this answers is not in the document]
Comment (ahg-g) [reply apis/kueue/v1alpha2/clusterqueue_types.go:298]: Shouldn't we support that under both LowerPriority (preempt lower priority and borrow the remaining) and Any (preempt any and borrow the remaining)?

[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/kueue/v1alpha2/clusterqueue_types.go:298]: Are you saying "preempt any|low-priority from ClusterQueue or and borrow the remaining"? This doesn't affect any workload in the cohort. We might be able to support this in the near future without changing the API. It's a bit of an optimization and I wouldn't call this "stealing".

These two cases, that I call "stealing", require an explicit API IMO:
- preempt low priority workloads in the cohort that are currently borrowing, but the action leads to the preempting ClusterQueue to borrow.
- (a bit more extreme) preempt low priority workloads in the cohort from ClusterQueues that are not currently borrowing

[the remark this answers is not in the document]
Comment (mwielgus) [reply apis/kueue/v1alpha2/clusterqueue_types.go:298]: If I may suggest, I think the constants should be the same for both preemption on borrowing and preemption on reclaiming so I like the proposal @alculquicondor made 3 comments above.

[the remark this answers is not in the document]
Comment (ahg-g) [reply apis/kueue/v1alpha2/clusterqueue_types.go:298]: > An alternative:
> 
> ```yaml
> preemption:
>   reclaimWithinCohort: Never|LowerPriority|Any
>   withinClusterQueue: Never|LowerPriority
> ```

isn't that what I suggested?

[the remark this answers is not in the document]
Comment (ahg-g) [reply apis/kueue/v1alpha2/clusterqueue_types.go:298]: ok, you added reclaim prefix, sure.

File: test/integration/webhook/v1alpha2/clusterqueue_test.go [hunk 5]
```diff
@@ -116,50 +186,76 @@ var _ = ginkgo.Describe("ClusterQueue Webhook", func() {
 				gomega.Expect(err).Should(gomega.HaveOccurred())
 				gomega.Expect(errors.IsInvalid(err)).Should(gomega.BeTrue(), "error: %v", err)
 			default:
-				gomega.Expect(err).Should(gomega.Succeed())
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply test/integration/webhook/v1alpha2/clusterqueue_test.go:119]: This is the alternative writing: `gomega.Expect(err).ShouldNot(gomega.HaveOccurred())`
Just being idiomatic.

File: apis/kueue/v1alpha2/clusterqueue_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/kueue/v1alpha2/clusterqueue_types.go]: Updated

File: pkg/scheduler/preemption/preemption_test.go [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/constants"
+	"sigs.k8s.io/kueue/pkg/scheduler/flavorassigner"
+	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+func TestPreemption(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/preemption/preemption_test.go:44]: 95%

File: pkg/scheduler/preemption/preemption_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/preemption/preemption_test.go:90]: Oh, I missed this comment. That's a good test case, I'll add one.

File: test/integration/scheduler/preemption_test.go (hunk 2 above)
Comment (alculquicondor) [reply test/integration/scheduler/preemption_test.go:84]: #530

File: pkg/scheduler/preemption/preemption.go (hunk 3 above)
Comment (tenzen-y) [reply pkg/scheduler/preemption/preemption.go:44]: Sounds good.

File: pkg/scheduler/preemption/preemption_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/preemption/preemption_test.go:90]: Added

File: apis/kueue/v1alpha2/clusterqueue_types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/kueue/v1alpha2/clusterqueue_types.go]: > Are you saying "preempt any|low-priority from ClusterQueue and borrow the remaining"? This doesn't affect any workload in the cohort. We might be able to support this in the near future without changing the API. It's a bit of an optimization and I wouldn't call this "stealing".

tracking issue #533

File: apis/kueue/webhooks/clusterqueue_webhook.go [hunk 7]
```diff
@@ -38,13 +38,16 @@ const (
 	isNegativeErrorMsg string = `must be greater than or equal to 0`
 )
 
-type ClusterQueueWebhook struct{}
+type ClusterQueueWebhook struct {
+	scheme *runtime.Scheme
```
Comment (ahg-g) [inline apis/kueue/webhooks/clusterqueue_webhook.go:42]: where is this used?

**Review by ahg-g — COMMENTED** [review]
/lgtm
/hold

for one nit

File: apis/kueue/webhooks/clusterqueue_webhook.go (hunk 7 above)
Comment (alculquicondor) [reply apis/kueue/webhooks/clusterqueue_webhook.go]: Left over from another approach I tried. Removed

**Comment by alculquicondor** [comment]
squashed into 2 commits

**Pull request merged**